### PR TITLE
[IMP] payment: improve allow unarchiving the archived tokens

### PR DESCRIPTION
**Pull request by rupa-odoo**
Before this commit,
When a payment method or payment provider was deactivated the tokens linked to it were also getting archived and they were prohibited from being active again.

After this commit, the tokens can be unarchived but if the linked pm and the pp if both are active.

task-3756355

### CODE REVIEW

File: addons/payment/models/payment_token.py [hunk 1]
```diff
@@ -83,8 +83,9 @@ def write(self, values):
         """
         if 'active' in values:
             if values['active']:
-                if any(not token.active for token in self):
-                    raise UserError(_("A token cannot be unarchived once it has been archived."))
+                for token in self:
+                    if not(token.payment_method_id.active and not token.provider_id.state == "disabled"):
+                        raise UserError(_("Once you archive the payment method and provider, you can't unarchive the token."))
```
Comment (kcv-odoo) [inline addons/payment/models/payment_token.py:88]: Why did you remove any? :thinking:
Also you can move you `not` operator inside to make condition little simpler. No?

File: addons/payment/tests/test_payment_token.py [hunk 2]
```diff
@@ -31,6 +31,8 @@ def test_cannot_assign_token_to_public_partner(self):
     def test_token_cannot_be_unarchived(self):
         """ Test that unarchiving disabled tokens is forbidden. """
         token = self._create_token(active=False)
+        token.payment_method_id.active = False
+        token.provider_id.state='disabled'
```
Comment (kcv-odoo) [inline addons/payment/tests/test_payment_token.py:35]: May be add other assertion with provider enable or not disable and other with method active or not

**Review by kcv-odoo — CHANGES_REQUESTED** [review]
Thank for you work! :100: 
Added 2 comments to look into it

File: addons/payment/models/payment_token.py (hunk 1 above)
Comment (rupa-odoo) [reply addons/payment/models/payment_token.py]: Done the changes :+1:

File: addons/payment/tests/test_payment_token.py (hunk 2 above)
Comment (rupa-odoo) [reply addons/payment/tests/test_payment_token.py]: Hello @kcv-odoo ,
I have added assertion with payment provider and method.

Thankyou :smile:

File: addons/payment/models/payment_token.py [hunk 3]
```diff
@@ -83,8 +83,8 @@ def write(self, values):
         """
         if 'active' in values:
             if values['active']:
-                if any(not token.active for token in self):
-                    raise UserError(_("A token cannot be unarchived once it has been archived."))
+                if any(not token.payment_method_id.active or token.provider_id.state == "disabled" for token in self):
```
Comment (kcv-odoo) [inline addons/payment/models/payment_token.py:86]: Nitpicking! I think your lines are more then 100 char :thinking:

**Review by kcv-odoo — APPROVED** [review]

File: addons/payment/models/payment_token.py [hunk 4]
```diff
@@ -83,8 +83,14 @@ def write(self, values):
...
             if values['active']:
-                if any(not token.active for token in self):
-                    raise UserError(_("A token cannot be unarchived once it has been archived."))
+                if any(
+                    not token.payment_method_id.active
+                    or token.provider_id.state == "disabled"
+                    for token in self
+                ):
+                    raise UserError(
+                        _("Once you archive the payment method and provider, you can't unarchive the token.")
```
Comment (kcv-odoo) [inline addons/payment/models/payment_token.py:92]: I think message is wrong :(
it should be
`You can't unarchive token linked to inactive payment method or disable provider`
@dit-odoo could you validate error message? :smiley:

Comment (dit-odoo) [reply addons/payment/models/payment_token.py:92]: @kcv-odoo

Hello,
For provider:  " Disabling the provider will also archive 2 tokens associated with it " 
For tokens: ok with what you suggest, "You can't unarchive tokens linked to inactive payment methods or disabled providers"

Comment (rupa-odoo) [reply addons/payment/models/payment_token.py]: Done the changes :+1: 

Thankyou so much :smile:

File: addons/payment/tests/test_payment_token.py [hunk 5]
```diff
@@ -31,6 +31,11 @@ def test_cannot_assign_token_to_public_partner(self):
     def test_token_cannot_be_unarchived(self):
         """ Test that unarchiving disabled tokens is forbidden. """
```
Comment (AntoineVDV) [inline addons/payment/tests/test_payment_token.py:32]: The method name and its docstring should be updated.
```suggestion
    def test_unarchiving_token_requires_active_provider(self):
        """ Test that unarchiving disabled tokens is forbidden if the provider is disabled. """
```

File: addons/payment/tests/test_payment_token.py [hunk 6]
```diff
@@ -31,6 +31,11 @@ def test_cannot_assign_token_to_public_partner(self):
...
         token = self._create_token(active=False)
+        token.payment_method_id.active = False
```
Comment (AntoineVDV) [inline addons/payment/tests/test_payment_token.py:34]: Please extract this case in a separate `test_unarchiving_token_requires_active_payment_method` test to keep investigation and maintenance efforts low.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hello, please find my review below :)

File: addons/payment/tests/test_payment_token.py (hunk 6 above)
Comment (rupa-odoo) [reply addons/payment/tests/test_payment_token.py:41]: Done the changes :+1:

File: addons/payment/models/payment_method.py [hunk 7]
```diff
@@ -131,8 +131,8 @@ def _onchange_warn_before_disabling_tokens(self):
                     'warning': {
                         'title': _("Warning"),
                         'message': _(
-                            "This action will also archive %s tokens that are registered with this "
-                            "payment method. Archiving tokens is irreversible.", len(related_tokens)
+                            "Are you sure that you want to archive this record? It will also archive the %s tokens linked to this payment method."
```
[the remark this answers is not in the document]
Comment (AntoineVDV) [reply addons/payment/models/payment_method.py:134]: Resolved without change.

[the remark this answers is not in the document]
Comment (AntoineVDV) [reply addons/payment/models/payment_method.py:134]: Resolved without change.

File: addons/payment/tests/test_payment_token.py [hunk 8]
```diff
@@ -28,9 +28,21 @@ def test_cannot_assign_token_to_public_partner(self):
...
+            token.active = True
+
+    def test_unarchiving_token_requires_active_payment_method(self):
+        """ Test that unarchiving disabled tokens is forbidden if the method is disabled. """
+        token = self._create_token(active=False)
+        token.payment_method_id.active = False
+        token.provider_id.state = 'enabled'
+        with self.assertRaises(UserError):
+            token.active = True
+        token.provider_id.state = 'disabled'
```
Comment (AntoineVDV) [inline addons/payment/tests/test_payment_token.py:45]: No need to test this behavior based on the provider state as it's already done with `test_unarchiving_token_requires_active_provider`.
```suggestion
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment/models/payment_method.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rupa-odoo) [reply addons/payment/models/payment_method.py]: Done

**Review by AntoineVDV — APPROVED** [review]
I think we're good to go :)

@robodoo r+
